torchao/prototype/quantized_training/bitnet.py

```diff
@@ -0,0 +1,345 @@
+        if isinstance(p, DTensor) and isinstance(p._local_tensor, BitNetTrainingLinearWeight)
+    ]
+    if len(bitnet_params) == 0:
+        return
+
+    # NOTE: use torch.compile to save memory and increase speed?
+    bitnet_scales = [get_bitnet_scale(x) for x in bitnet_params]  # local absmean
+    bitnet_scales = torch.stack(bitnet_scales)
+    bitnet_scales = bitnet_scales.full_tensor()  # global absmean
+
+    for i, p in enumerate(bitnet_params):
+        p._local_tensor._precomputed_scale = bitnet_scales[i]
+
+
+class _BitNetTrainingLinear(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, input: Tensor, weight: BitNetTrainingLinearWeight, bias: Optional[Tensor] = None):
+        batch_dims = input.shape[:-1]
+        input = input.view(-1, weight.shape[1])
+
+        # https://github.com/microsoft/unilm/blob/master/bitnet/The-Era-of-1-bit-LLMs__Training_Tips_Code_FAQ.pdf
+        # Figure 3
+        input_i8, row_scale = quantize_int8_rowwise(input, eps=1e-5)
+
+        # NOTE: use FP32 scale for weight quantization, but cast scale to possibly lower precision
+        # for matmul and backward
+        tensor_scale = get_bitnet_scale(weight._data)
+        weight_i8 = quantize_bitnet_weight(weight._data, tensor_scale)
+        tensor_scale = tensor_scale.to(weight.dtype)
+
+        ctx.save_for_backward(input_i8, row_scale, weight_i8, tensor_scale)
+
+        # use int8 tensor cores
+        out = scaled_int8_mm(input_i8.contiguous(), weight_i8.contiguous().T, row_scale, tensor_scale)
+        out = out.view(*batch_dims, weight.shape[0])
+
+        out = out + bias if bias is not None else out
+        return out
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        input_i8, row_scale, weight_i8, tensor_scale = ctx.saved_tensors
+        grad_input = grad_weight = grad_bias = None
+
+        batch_dims = grad_output.shape[:-1]
+        grad_output = grad_output.view(-1, weight_i8.shape[0])
+
+        # NOTE: we can potentially speedup training by also quantizing the backward pass
+        # to use INT8 tensor cores
+        if ctx.needs_input_grad[0]:
+            # mixed mm
+            grad_input = (grad_output @ weight_i8.to(grad_output.dtype)) * tensor_scale
+            grad_input = grad_input.view(*batch_dims, weight_i8.shape[1])
+
+        if ctx.needs_input_grad[1]:
+            # NOTE: we use quantized activation for this calculation
+            grad_weight = grad_output.T @ (input_i8 * row_scale.view(-1, 1))
+
+        if ctx.needs_input_grad[2]:
+            grad_bias = grad_output.sum(0)
+
+        return grad_input, grad_weight, grad_bias
+
+
+def bitnet_training():
+    return _get_linear_subclass_inserter(BitNetTrainingLinearWeight, allow_requires_grad=True)
+
+
+def _pack_i2_in_i8(x: Tensor):
+    # NOTE: this is signed integer, so we have to mask before bit-shift
+    return (x[:, ::4] << 6) | ((x[:, 1::4] & 0b11) << 4) | ((x[:, 2::4] & 0b11) << 2) | (x[:, 3::4] & 0b11)
```

Comment:
readability nit: write it out line by line with comments to make easier to understand for code readers?